appsrc/dcfile/dchist.man
.TH DCHIST 1 "20 August 2007" "DICOM PS3" "DICOM PS3 - DICOM image statistics"
.SH NAME
dchist \- ACR/NEMA DICOM PS3 ... DICOM PS3 - DICOM image statistics
.SH SYNOPSIS
.HP 10
.B dchist
[
.B \-h|histogram
]
[
.B \-csv
]
[
.B \-withhighbits|highbitsonly
]
[
.B \-v|verbose
]
[
.B \-filename
]
.so man1/optin.so
.SH DESCRIPTION
.LP
.B dchist
reads the named dicom or acr-nema input file and describes the statistics
of the image pixel data. The primary intent is to determine the zero
order entropy of the image.
.LP
The description and verbose output go to standard error.
.PP
The basic input switches are described in dcintro(1). Options specific to this program are:
.TP
.B \-h|histogram
.RS
Display table of pixel value, frequency, probability, contribution to zero order entropy, and cumulative zero order entropy. May generate a LOT of output !
.RE
.TP
.B \-csv
.RS
Display only a comma separated value pair of pixel value and frequency, one line per value, e.g., for import into a statistical program or spreadsheet
.RE
.TP
.B \-withhighbits
.RS
Include as well the values that are above Bits Stored and within Bits Allocated, such as might be overlay bits
(regardless of whether or not there are any 60xx group elements).
.RE
.TP
.B \-highbitsonly
.RS
Use instead the values that are above Bits Stored and within Bits Allocated, such as might be overlay bits,
and treat all valid pixel values below Bits Stored inclusive as zero
(regardless of whether or not there are any 60xx group elements).
.RE
.TP
.B \-v|verbose
.RS
Display byte offset from file start and message start, in hexadecimal, and dump contents while reading and once read.
.RE
.TP
.B \-filename
.RS
Show the name of the file supplied in the arguments; a hyphen will be reported if no filename was supplied.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
\ 
.RE
% dchist expmeta.dc3
.RE
9.02366 Zero order entropy (bits per pixel)
.RE
0x0     Smallest symbol value
.RE
0       Smallest symbol length (bits)
.RE
0x6e9   Largest symbol value
.RE
11      Largest symbol length (bits)
.RE
\ 
.RE
% dchist -h expmeta.dc3
.RE
[0x0]   392 p=0.00598145  e=0.0441747   cum=0.0441747
.RE
 ...
.RE
[0x6e9] 1   p=1.52588e-05 e=0.000244141 cum=9.02366
.RE
\ 
.RE
9.02366 Zero order entropy (bits per pixel)
.RE
0x0     Smallest symbol value
.RE
0       Smallest symbol length (bits)
.RE
0x6e9   Largest symbol value
.RE
11      Largest symbol length (bits)
.RE
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR dcstats(1) ,
.BR dcintro(1)
.SH AUTHOR
Copyright (C) 1993-2010. David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS


CHANGES
111219:	libsrc/src/dctool/attrmxrd.cc: check for HighBit being more than one less than BitsStored and correct BitsStored accordingly (otherwise will create zero pixel values on copy) (000408)

111213:	appsrc/dcfile/dcuncat.cc: add option to replace instance number starting with a specific number to allow scripts to extend a series with successive invocations

111212:	libsrc/standard/condn.tpl/module,strval/base.tpl,dx.tpl: improve animal versus human conditions to check SRT homo sapiens, and make view position and body part examined human versus animal specific

111212:	libsrc/standard/elmdict/other.tpl: add more Sound private elements

111117:	libsrc/standard/strval/base.tpl: add CSD for 99NCIAIM

111109:	appsrc/dcfile/dcuncat.cc: add options to replace instance number, frame increment pointer and frame increment pointer vectors, as well as add number of frames during unenhance based on SOP class

111106:	libsrc/standard/elmdict/other.tpl: add Vepro private elements

111031:	libsrc/standard/elmdict/other.tpl: add CTP private elements

111030:	libsrc/standard/elmdict/other.tpl: add Segami private elements

111029:	libsrc/standard/elmdict/dicom3.tpl: add (restore from trial) Observation UID (per CP 1147)

111028:	libsrc/standard/elmdict/other.tpl: add NeuroQuant private elements

111024:	libsrc/standard/elmdict/other.tpl: update fuji private attributes with descriptions

110918:	libsrc/standard/elmdict/diconde.tpl: update with new or modified data elements with cleanup from Patrick

110918:	libsrc/standard/elmdict/siemens.tpl: update some private elements based on Artis conformance statement

110916:	libsrc/standard/elmdict/gems.tpl: resolve inconsistencies in GE PET conformance statement VRs SS/SL/FL for private elements by choosing those most commonly seen in the field (still some inconsistencies between models)

110904:	libsrc/standard/elmdict/camtron.tpl: add private data elements encountered in DICOMDIR, as well as QCA private data elements from GEMS/Camtronics CRS conformance statement

110901:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Sup 152 FT ophthalmic thickness map (dictionary and SOP class only; not IOD yet)

110822:	libsrc/standard/elmdict/dicom3.tpl: update data dictionary (only) for CPs for FT Aug 2011 WG 6 meeting

110714:	libsrc/standard/elmdict/dicom3.tpl: remove 'unified' from names of various attributes (per Sup 96 FT)

110716:	libsrc/standard/strval/base.tpl: add 99PMP CSD

110714:	libsrc/standard/elmdict/dicom3.tpl: exposure index etc are DS not FD (CP 1024)

110709:	libsrc/standard/elmdict/dicom3.tpl: change (0018,9430) from US to FL per Sup 139

110612:	libsrc/standard/elmdict/diconde.tpl: update for DICONDE eddy current standard

110612:	libsrc/standard/elmdict/dicondep.tpl: correct case of DICONDE DX (Dx) private owner for E2699

110609:	libsrc/standard/elmdict/dicom3.tpl: add Sup 151 FT data elements

110608:	libsrc/standard/elmdict/elscint.tpl: update Elscint private elements

110531:	libsrc/standard/elmdict/other.tpl: update Hitachi CT private elements

110531:	libsrc/standard/elmdict/siemens.tpl: update mammo private elements

110529:	libsrc/standard/condn.tpl,iodcomp/dx.tpl,module/base.tpl,dx.tpl,xaxrf.tpl,strval/dx.tpl,xaxrf.tpl: add breast tomosynthesis IOD, refactoring 3D XRay accordingly

110529:	libsrc/standard/elmdict/other.tpl: more private data elements for Hologic 

110528:	appsrc/dcfile/dciodvfy.cc: do not complain about SCOORD without child IMAGE when SELECTED FROM is by reference (000406)

110527:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: check that references to instances in content tree are in top level evidence sequences in SR (000405)

110526:	libsrc/standard/elmdict/gems.tpl: add names for CT iterative reconstruction private data elements

110522:	libsrc/src/dctool/attrothr.cc: do not throw assertion if computed length is equal (which may occur for single bit images)

110522:	libsrc/standard/iodcomp/base.tpl,libsrc/standard/condn.tpl: permit FoR Module in Segementation (and do CP to standard to allow may be present otherwise)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect SCOORD without child IMAGE (000403)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect quotes in code meaning of measurement units (000404)

110522:	libsrc/src/dctool/attrtypv.cc,locale/mesgtext.tpl: permit PN with multiple carets in multiple component groups (000402)

110521:	libsrc/standard/elmdict/acuson.tpl,agfa.tpl,isg.tpl,other.tpl,siemens.tpl,strval/base.tpl: more private elements and coding scheme designators

110517:	libsrc/standard/module/base.tpl: correct type (1->2) and number of items (1->0-n) in SourceImageSequence in DerivationImageMacro

110517:	libsrc/standard/elmdict/siemens.tpl: yet more syngo private elements

110516:	libsrc/standard/cond.tpl: no longer require version for UCUM codes (CP 1060)

110516:	libsrc/standard/strval/base.tpl: add 99_MI private coding scheme designator

110516:	libsrc/include/dctool/attr.h,attrseq.h,generic/listsimp.h,src/dctool/attrseq.cc,attrvrfy.cc,standard/condn.tpl,module/softcopy.tpl,sr.tpl,support/module.awk:
		though VM for sequences number of items was being specified in modules, it was not being applied;
		apply it (000256), as well as make it work (since number of items was not being computed correctly
		in SequenceAttribute::verifyVM()); also add support for conditional check of VM, for example for
		GraphicData depending on Grahic Type (000401)

110516:	libsrc/standard/elmdict/toshiba.tpl: update CT private elements from conformance statement for V4.70 or later version

110514:	libsrc/standard/elmdict/siemens.tpl: add more syngo private elements, and update guessed VRs of existing ones with observed explicit VRs

110511:	libsrc/standard/elmdict/gems.tpl,toshiba.tpl: add more encountered explicit VR private elements

110511:	libsrc/standard/elmdict/other.tpl: update/add McKesson private elements from conformance statement


appsrc/dcfile/dchist.cc
#include "attrmxls.h"
#include "attrothr.h"
#include "attrval.h"
#include "elmconst.h"
#include "mesgtext.h"
#include "dcopt.h"

static unsigned
howManySignificantBits(Uint16 test)
{
	unsigned bits;
	for (bits=0; test; test=test>>1,++bits);
	return bits;
}

static bool
dumpImageHistogram(ManagedAttributeList& list,TextOutputStream& log,bool showhist,bool showcsv,bool withHighBits,bool highBitsOnly)
{
	bool success;

	Attribute *aBitsStored = list[TagFromName(BitsStored)];
	Uint16 vBitsStored=AttributeValue(aBitsStored);

	Attribute *aBitsAllocated = list[TagFromName(BitsAllocated)];
	Uint16 vBitsAllocated=AttributeValue(aBitsAllocated);

	Attribute *aRows = list[TagFromName(Rows)];
	if (!aRows)
		log << EMsgDC(MissingAttribute) << " - \"Rows\"" << endl;
	Uint16 vRows=AttributeValue(aRows);

	Attribute *aColumns = list[TagFromName(Columns)];
	if (!aColumns)
		log << EMsgDC(MissingAttribute) << " - \"Columns\"" << endl;
	Uint16 vColumns=AttributeValue(aColumns);

	Attribute *aNumberOfFrames = list[TagFromName(NumberOfFrames)];
	Uint16 vNumberOfFrames=AttributeValue(aNumberOfFrames,1);

	Attribute *aPixelData = list[TagFromName(PixelData)];
	if (!aPixelData)
		log << EMsgDC(MissingAttribute) << " - \"PixelData\"" << endl;

	if (!aPixelData || !aRows || !aColumns) {
		success=false;
	}
	else {
		SupplySourceFromAttribute sPixelData(aPixelData,withHighBits || highBitsOnly);
		class SourceBase<Uint16> *source = sPixelData.getSource();
		if (!source) {
			log << EMsgDC(BadAttributeValue)
			    << " - \"PixelData\""
			    << endl;
			success=false;
		}
		else {
			int bitsUsed;
			Uint16 maskLowBits = 0;
			if (withHighBits) {
				bitsUsed = vBitsAllocated;
			}
			else if (highBitsOnly) {
				bitsUsed = vBitsAllocated;
				int i;
				for (i=0; i<vBitsStored; ++i) {
					maskLowBits = (maskLowBits<<1) | 1;
				}
			}
			else {
				bitsUsed = vBitsStored;
			}
			maskLowBits = (~maskLowBits)&0xffff;
			Uint32 largestsymbolplusone = ((Uint32)1) << (bitsUsed ? bitsUsed : 16);
			Uint32 *histogram = new Uint32[largestsymbolplusone];
			Assert(histogram);
			{
				Uint32 i;
				Uint32 *histptr;
				for (i=0,histptr=histogram; i<largestsymbolplusone; ++i) *histptr++=0;
			}
			Uint32 count=0;
			Uint16 smallestsymbol=0xffff;
			Uint16 largestsymbol=0;

			unsigned row=0;
			unsigned column=0;
			while (row < vRows*vNumberOfFrames && source->read()) {
				size_t n=source->getBufferCount();
				Assert(n);
				const Uint16 *buffer=source->getBuffer();
				Assert(buffer);
				while (n-- && row < vRows*vNumberOfFrames) {
					Uint16 symbol=*buffer++;
					symbol = symbol & maskLowBits;

					Assert(symbol<largestsymbolplusone);
					Assert(histogram[symbol] < 0xffffffff);
					++histogram[symbol];
					++count;

					if (symbol < smallestsymbol) smallestsymbol=symbol;
					if (symbol > largestsymbol) largestsymbol=symbol;

					if (++column >= vColumns) {
						column=0;
						++row;
					}
				}
			}
			//Assert(column == 0 && row == vRows*vNumberOfFrames);

			Assert(count == vColumns*vRows*vNumberOfFrames);
			Assert(count);

			double entropy;
			{
				Uint32 i;
				Uint32 *histptr;
				for (i=smallestsymbol,histptr=histogram+smallestsymbol,entropy=0; i<=largestsymbol; ++i,++histptr) {
					if (*histptr) {
						double probsymbol=double(*histptr)/count;
						double contributes=-probsymbol*ourlog2(probsymbol);
						entropy+=contributes;
						if (showhist) {
							log << "[" << hex << i << "]\t"
							    << dec << *histptr << "\t"
							    << "p=" << dec << probsymbol << "\t"
							    << "e=" << dec << contributes << "\t"
							    << "cum=" << dec << entropy << endl;
						}
						if (showcsv) {
							log << dec << i << "," << *histptr << endl;
						}
					}
				}
			}

			if (showhist) log << endl;

			if (!showcsv) {
				log << dec << entropy << "\tZero order entropy (bits per pixel)" << endl;
				log << hex << smallestsymbol << "\tSmallest symbol value" << endl;
				log << dec << howManySignificantBits(smallestsymbol) << "\tSmallest symbol length (bits)" << endl;
				log << hex << largestsymbol << "\tLargest symbol value" << endl;
				int mostSignificantBits = howManySignificantBits(largestsymbol);
				log << dec << mostSignificantBits << "\tLargest symbol length (bits)" << endl;
				if (mostSignificantBits > vBitsStored) {
					log << "ALERT\tEmbedded overlay bits or sign extension bits present above Bits Stored" << endl;
				}
			}
			
			success=true;
		}
	}
	return success;
}

int
main(int argc, char *argv[])
{
	GetNamedOptions 	options(argc,argv);
	DicomInputOptions 	dicom_input_options(options);

	bool verbose=options.get("verbose") || options.get("v");
	bool showfilename=options.get("filename");
	bool showhist=options.get("histogram") || options.get("h");
	bool showcsv=options.get("csv");
	bool withHighBits=options.get("withhighbits");
	bool highBitsOnly=options.get("highbitsonly");
	
	dicom_input_options.done();
	options.done();

	DicomInputOpenerFromOptions input_opener(
		options,dicom_input_options.filename,cin);

	cerr << dicom_input_options.errors();
	cerr << options.errors();
	cerr << input_opener.errors();

	if (!dicom_input_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !options) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< dicom_input_options.usage()
			<< " [-h|-histogram]"
			<< " [-csv]"
			<< " [-v|-verbose]"
			<< " [-withhighbits|-highbitsonly]"
			<< " [-filename]"
			<< " [" << MMsgDC(InputFile) << "]"
			<< " <" << MMsgDC(InputFile)
			<< endl;
		exit(1);
	}

	DicomInputStream din(*(istream *)input_opener,
		dicom_input_options.transfersyntaxuid,
		dicom_input_options.usemetaheader);

	ManagedAttributeList list;

	bool success=true;
	TextOutputStream log(cerr);

	if (showfilename) {
		const char *filenameused = input_opener.getFilename();
		log << "Filename: \"" << (filenameused && strlen(filenameused) > 0 ? filenameused : "-") << "\"" << endl;
	}

	if (verbose) log << "******** While reading ... ********" << endl; 
	list.read(din,&log,verbose,0xffffffff,true,dicom_input_options.uselengthtoend,dicom_input_options.ignoreoutofordertags,dicom_input_options.useUSVRForLUTDataIfNotExplicit);

	const char *errors=list.errors();
	if (errors) log << errors << flush;
	if (!list.good()) {
		log << EMsgDC(DatasetReadFailed) << endl;
		success=false;
	}
	else {
		success=dumpImageHistogram(list,log,showhist,showcsv,withHighBits,highBitsOnly);
	}

	return success ? 0 : 1;
}

